### PR TITLE
Validate origin format based on issue tracker

### PR DESCRIPTION
**Pull request by srafi1**
Solves #25

### CODE REVIEW

**Comment by srafi1** [comment]
I'm not sure how to validate the origin for Jira or Redmine, but this should validate the others properly

File: config/config.go
Comment (preslavmihaylov) [inline config/config.go:85]: Could you convert this switch-case into a map - `map[IssueTracker]*Regexp`.

Afterwards, validation can look like:
```
pattern, !ok := originPatterns[cfg.IssueTracker]
if !ok {
   // return err
}

// validate pattern matches
```

File: config/config.go
Comment (preslavmihaylov) [inline config/config.go:67]: Could you pull latest master & move this chunk into the `Validate` function for consistency? 

See #46

File: config/config.go
Comment (preslavmihaylov) [inline config/config.go:36]: For redmine, the regex should match any domain as it is self-hosted (e.g. `https://my-redmine-instance.com`)

There is no expected path appended to the hosts as redmine issues are accessed as `https://my-redmine-instance.com/issues/3`

Comment (preslavmihaylov) [reply config/config.go:36]: You can test integration \w Redmine by signing up to their demo hosted website - http://demo.redmine.org/

File: config/config.go
Comment (preslavmihaylov) [inline config/config.go:31]: For Jira, the origin is similar to redmine. It can be any domain (e.g. `https://myjira.com`).

File: config/config.go
Comment (preslavmihaylov) [reply config/config.go:85]: Also, please make sure this map is close to the enum declaration to make things easier for future maintainers once a new integration is added.

**Review by preslavmihaylov — CHANGES_REQUESTED** [review]
Thanks for taking the time to put together this PR!

Left some comments & suggestions for validating Jira and redmine.

I can help with testing the new validations for all supported issue trackers once you're ready.

File: config/config.go
Comment (srafi1) [reply config/config.go]: resolved in eaf4e4a

File: config/config.go
Comment (srafi1) [reply config/config.go]: resolved in 78a7217

File: config/config.go
Comment (srafi1) [reply config/config.go]: resolved in accb53f

File: config/config.go
Comment (srafi1) [reply config/config.go]: resolved in accb53f

**Comment by srafi1** [comment]
Thanks for the comments. I fixed everything you mentioned, but I have a couple of questions. Why does `Validate()` return a list of errors when it can only ever have one according to the current logic? Also, when I ran the tests with `go test ./testing -v`, many of the tests failed. I ended up running my own tests using `go test ./testing -run Origins -v` and although they pass, I'm not sure if I'm running tests properly if the other ones are failing on my computer.

File: config/config.go
Comment (preslavmihaylov) [inline config/config.go:44]: This pattern excludes valid origins (i.e subdomains) such as `https://redmine.myorg.com`.

Please extend the pattern to allow such origins & add a test case for that.

File: config/config.go
Comment (preslavmihaylov) [inline config/config.go:40]: This pattern excludes valid origins (i.e subdomains) such as `https://myjira.myorg.com`.

Please extend the pattern to allow such origins & add a test case for that.

File: config/config.go
Comment (preslavmihaylov) [inline config/config.go:42]: Some companies have a private instance of gitlab setup. E.g. `https://mygitlab.org.com`.

`todocheck` has support for private gitlab instances but this validation excludes them.

Example - #40 

This pattern should (similarly to jira & redmine) match any (sub)domain.

File: testing/test_configs/valid_github_origin.yaml
Comment (preslavmihaylov) [inline testing/test_configs/valid_github_origin.yaml:1]: Could you also add valid configs for all supported issue trackers:
 * Without a scheme provided - e.g. `github.com/user/project`
 * With `www` provided - e.g. `www.github.com/user/project` and `https://www.github.com/user/project`

File: testing/todocheck_test.go
Comment (preslavmihaylov) [inline testing/todocheck_test.go:362]: This is a pretty neat way to test all these.
It avoids adding a separate test case for all issue trackers.

Thanks!

**Review by preslavmihaylov — CHANGES_REQUESTED** [review]
I have a few more comments left to make this fully compatible with what's currently supported & for adding more extensive test cases.

**Comment by preslavmihaylov** [comment]
> Thanks for the comments. I fixed everything you mentioned, but I have a couple of questions. Why does `Validate()` return a list of errors when it can only ever have one according to the current logic? Also, when I ran the tests with `go test ./testing -v`, many of the tests failed. I ended up running my own tests using `go test ./testing -run Origins -v` and although they pass, I'm not sure if I'm running tests properly if the other ones are failing on my computer.

The tests all pass when I run them from latest master:
image

Could you try running the test suite from latest master & see if they pass?

As for the `Validate` function, the goal is to batch report all issues with the configuration. 

Currently, your validation is incompatible with the rest but as more validations are added, there will be independent issues \w configurations which can be reported to the user.

File: config/config.go
Comment (preslavmihaylov) [inline config/config.go:93]: This check is dependent on the order of validations. This shouldn't happen as this piece of validation might as well become first inline in the future.

Instead, change your condition to be independent from the `errors` array by validating that the issue tracker is not empty:
```
if l.IssueTracker != "" {
   ...
}
```

**Comment by preslavmihaylov** [comment]
Blocked by #54

**Comment by preslavmihaylov** [comment]
Hey, the PR blocking this one is now merged. Please pull latest master & continue from there (to enable the CI checks).

File: config/config.go
Comment (srafi1) [reply config/config.go]: fixed in f4977d0 with test case in defa509

File: config/config.go
Comment (srafi1) [reply config/config.go]: fixed in f4977d0 with test case in defa509

File: config/config.go
Comment (srafi1) [reply config/config.go]: fixed in 56054bf and 4b08267 with test case in defa509

File: testing/test_configs/valid_github_origin.yaml
Comment (srafi1) [reply testing/test_configs/valid_github_origin.yaml]: added in defa509

File: config/config.go
Comment (srafi1) [reply config/config.go]: fixed in f4977d0

**Comment by srafi1** [comment]
I figured out the testing issue on my end, added a bunch of tests for the regex, and fixed the regex according to your comments. LMK if there's anything else to add.

File: config/config.go
Comment (preslavmihaylov) [inline config/config.go:159]: Could you make the err message slightly more informative by writing:
```
fmt.Sprintf("%s is not a valid origin for issue tracker %s", l.Origin, l.IssueTracker)
```

**Review by preslavmihaylov — CHANGES_REQUESTED** [review]
Everything looks good this time around in terms of implementation, thanks!

I only have a single comment for the err message format used to report issues with invalid origins.

File: config/config.go
Comment (srafi1) [reply config/config.go]: resolved in 16833af

**Review by preslavmihaylov — APPROVED** [review]

**Pull request merged**
